### ISSUE TITLE
Implement GadgetStrategy holding a &mut StandardComposer

### ISSUE THREAD

**Issue by CPerezz** [dusk-network/Hades252#46]
While building the examples of the gadget usage, I've realized that it's a bit weird how you're supposed to call the gadget functions exposed for the `GadgetStrategy`.

Actually, you need to copy the `StandardComposer` by value to the `hades_gadget()` function and then re-declare it with the mutated state.

This is not inline with the typical gadget function usage. 
The way of interacting with gadget functions should be to provide a `&mut StandardComposer` to the function so it gets modified (constraints get added to it) but you never need to re-declare the same object to be able to apply a gadget to it.

This is how it actually looks like:
```rust
 pub fn hades_gadget(
        composer: StandardComposer,
        pi: P,
        x: &mut [Variable],
    ) -> (StandardComposer, P) {
        #[cfg(feature = "trace")]
        let circuit_size = composer.circuit_size();

        let mut strategy = GadgetStrategy::new(composer, pi);

        strategy.perm(x);

        #[cfg(feature = "trace")]
        {
            trace!(
                "Hades permutation performed with {} constraints for {} bits",
                strategy.cs.circuit_size() - circuit_size,
                WIDTH
            );
        }

        strategy.into_inner()
    }
```

So how I envision this gadget function and the `GadgetStrategy` will look like:
```rust
/// Implements a Hades252 strategy for `Variable` as input values.
/// Requires a reference to a `ConstraintSystem`.
pub struct GadgetStrategy<'a> {
    /// A reference to the constraint system used by the gadgets
    pub cs: &'a mut StandardComposer,
}

impl<'a> GadgetStrategy<'a> {
    /// Constructs a new `GadgetStrategy` with the constraint system.
    /// In this case, the CS is embedded inside of a plonk `StandardComposer`.
    pub fn new(cs: &'a mut StandardComposer) -> Self {
        GadgetStrategy { cs }
    }

    /// Perform the hades permutation on a plonk circuit given a set of Public
    /// Inputs and a set of Variables used as inputs.
    pub fn hades_gadget(composer: &mut StandardComposer, inputs: &mut [Variable]) {
        #[cfg(feature = "trace")]
        let circuit_size = composer.circuit_size();

        let mut strategy = GadgetStrategy::new(&mut composer);

        strategy.perm(inputs);

        #[cfg(feature = "trace")]
        {
            trace!(
                "Hades permutation performed with {} constraints for {} bits",
                strategy.cs.circuit_size() - circuit_size,
                WIDTH
            );
        }
    }
//.........
}
```

On this way, we will be able to call the gadget on the way that is recommended by PLONK, see: https://github.com/dusk-network/plonk/blob/f2e889184961b885d9e095afe35b137165112c91/examples/3_1_final_gadget_orientation.rs#L49

The call will look like:
```rust
let mut composer = StandardComposer::new();
let mut inputs = [BlsScalar::one(); WIDTH];
GadgetStrategy::hades_gadget(&mut Composer, inputs);
// Now I have my gadget with the hades gadget printed into my internal ConstraintSystem.
```

**Comment by CPerezz** [comment]
WDYT @kevaundray @ZER0 ?

**Comment by CPerezz** [comment]
Closed via #47
